include/TPP/IR/StructuredOpMatcher.h

```diff
@@ -115,6 +115,26 @@ struct HasStaticShape {
   }
 };
 
+// Callable object to verify `operand` to have a rank in `ranks`.
+struct HasRank {
+  HasRank() = delete;
+  explicit HasRank(std::initializer_list<int64_t> ranks) : ranks(ranks){};
+
+  bool operator()(OpOperand *operand, Operation *op) const {
+    auto operandType = operand->get().getType();
+    if (!isa<ShapedType>(operandType))
+      return llvm::is_contained(ranks, HasRank::SCALAR);
+    int64_t rank = operandType.cast<ShapedType>().getRank();
+    return llvm::any_of(
+        ranks, [=](int64_t expectedRank) { return expectedRank == rank; });
+  }
+
+  // There are multiple way to represent a scalar: f32, tensor<f32>.
+  // SCALAR means f32.
```

Comment:
This is a weird requirement... 

"scalars" can be:
* `f32`, `bf16`, `i64` etc.
* `memref<f32>`, `memref<bf16>`, `memref<i64>` etc.
* `memref<1xf32>`, `memref<1xbf16>`, `memref<1xi64>` etc.
* `memref<1x1xf32>`, `memref<1x1xbf16>`, `memref<1x1xi64>` etc.
* etc

Reply:
I think if the input has form of `memref<1x1xf32>` etc. then it just falls under standard rank requirements i.e., 1D or 2D.
But it would be nice to support `memref<f32>`.

Reply:
indeed, memref<1x1xf32> is a 2d tensor. It does not make sense to handle this in the matchers.

Reply:
In other places of the code (broadcasting) we treat `1x1xf32` like a scalar.

We need to ask the question of what it the intention of this matcher: to match against the expectation or the literal. If it is the literal, we probably need another matcher to find the effective dimension (intention).

So, why is this matcher being added, and what are the surrounding matchers that will be changed/added/removed because of this?

Reply:
Exactly and we don't want to have to do that because bufferization may get it wrong. IF whatever comes from input as `memref<f32>` comes as an `f32`, then it should be fine to accept it as `0D`. 

This is why I'm asking: where is this coming from? What are you seeing that isn't shared in this PR?

Reply:
I am just trying to be pragmatic here. `tensor<1x1xf32>` has rank 2 even if it is a scalar. If we want to allow these kind of scalar we need an additional pass before to collapse. We should technically be able to handle both `f32` and `memref<f32`, the latter is not implemented but we can add support.